configuration/exodus-pipeline.yaml

```diff
@@ -0,0 +1,345 @@
+AWSTemplateFormatVersion: 2010-09-09
+
+Description: Configuration for exodus-cdn CodePipeline
+
+Parameters:
+  env:
+    Type: String
+    AllowedValues:
+      - dev
+      - stage
+      - prod
+    Default: dev
+    Description: The environment in which to deploy functions
+  oai:
+    Type: String
+    Description: The origin access identity ID associated with the environment
+  githubToken:
+    Type: String
+    Default: None
+    Description: A GitHub access token token to authorize webhooks
+  region:
+    Type: String
+    Default: us-east-1
+    Description: The region in which resources are established
+  email:
+    Type: String
+    Default: project-exodus@redhat.com
+    Description: The email address to which notifications are sent
+
+Conditions:
+  IsDev:
+    !Equals [!Ref env, dev]
+  NotDev:
+    !Not [!Equals [!Ref env, dev]]
+  NotProd:
+    !Not [!Equals [!Ref env, prod]]
+
+Mappings:
+  SourceMap:
+    dev:
+      objectKey: ""
+    stage:
+      objectKey: build-dev/exodus-lambda-pkg.yaml
+    prod:
+      objectKey: build-stage/exodus-lambda-pkg.yaml
+
+Resources:
+  SNSTopic:
+    Type: AWS::SNS::Topic
+    Properties:
+      TopicName: !Sub exodus-sns-${env}
+      Subscription:
+        - Endpoint: !Ref email
+          Protocol: email
+
+  CodePipelineNotification:
+    Type: AWS::CodeStarNotifications::NotificationRule
+    Properties:
+      Name: !Sub exodus-pipeline-${env}
+      DetailType: FULL
+      EventTypeIds:
+        - codepipeline-pipeline-pipeline-execution-failed
+      Resource:
+        !Sub arn:aws:codepipeline:${region}:${AWS::AccountId}:${CodePipeline}
+      Targets:
+        - TargetType: SNS
+          TargetAddress:
+            !Sub arn:aws:sns:${region}:${AWS::AccountId}:${SNSTopic}
+
+  #CodePipelineWebhook:
+  #  Type: AWS::CodePipeline::Webhook
+  #  Condition: IsDev
+  #  Properties:
+  #    Name: !Sub exodus-pipeline-${env}
+  #    AuthenticationConfiguration:
+  #      SecretToken: 123456789  # TODO
+  #    Filters:
+  #      - JsonPath: "$.ref"
+  #        MatchEquals: refs/heads/{Branch
+  #    Authentication: GITHUB_HMAC
+  #    TargetPipeline: !Ref CodePipeline
+  #    TargetAction: Source
+  #    TargetPipelineVersion: !GetAtt CodePipeline.Version
+  #    RegisterWithThirdParty: true
```

Comment:
@rohanpm I suspect we'll need this webhook setup, which will require GitHub settings I don't have access to.

Reply:
@nathanegillett I read the AWS docs relating to this.  Can you confirm my understanding of what's going on here:

- This resource causes AWS to create a webhook on Github, to call back to AWS when branches are changed.
- When authenticating with github to create this hook, AWS uses the OAuthToken from the Source action (i.e. "githubToken" parameter)
- That token needs to have "admin:repo_hook" scope (got this from https://docs.aws.amazon.com/codepipeline/latest/userguide/pipelines-webhooks-create-cfn.html).  That doc also implies it needs "repo" scope but I have a feeling it might not actually be correct for public repos.
- I guess you're using a token you created via your own account while working on this?

If that's all accurate: could you try using the token I just added to distribution-password-store named exodus-aws-oauth?  It has admin:repo_hook scope.  I did not add repo scope since, as I mentioned above, I think it might not actually be needed.

Reply:
That all sounds right. I've confirmed that only repo_hook is needed using a personal access token I generated and it works to create the webhook on my fork. I'll continue with this for now until the template is finalized; I'm still trying to work the CloudWatch alarms into deploy stage's CloudFormation for automatic rollbacks.

Thanks for adding the token to password-store. We can use that for actual pipeline deployments in the future.